pkg/services/accesscontrol/evaluator.go

```diff
@@ -0,0 +1,194 @@
+package accesscontrol
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+var logger = log.New("accesscontrol.evaluator")
+
+type Evaluator interface {
+	// Evaluate permissions that are grouped by action
+	Evaluate(permissions map[string]map[string]struct{}) (bool, error)
+	// Inject params into the evaluator's templated scopes. e.g. "settings:" + eval.Parameters(":id") and returns a new Evaluator
+	Inject(params map[string]string) (Evaluator, error)
+	// String returns a string representation of permission required by the evaluator
+	String() string
+}
+
+var _ Evaluator = new(permissionEvaluator)
+
+// EvalPermission returns an evaluator that will require all scopes in combination with action to match
+func EvalPermission(action string, scopes ...string) Evaluator {
+	return permissionEvaluator{Action: action, Scopes: scopes}
+}
+
+type permissionEvaluator struct {
+	Action string
+	Scopes []string
+}
+
+func (p permissionEvaluator) Evaluate(permissions map[string]map[string]struct{}) (bool, error) {
+	userScopes, ok := permissions[p.Action]
+	if !ok {
+		return false, nil
+	}
+
+	if len(p.Scopes) == 0 {
+		return true, nil
+	}
+
+	for _, target := range p.Scopes {
+		var err error
+		var matches bool
+
+		for scope := range userScopes {
+			matches, err = match(scope, target)
+			if err != nil {
+				return false, err
+			}
+			if matches {
+				break
+			}
+		}
+		if !matches {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func match(scope, target string) (bool, error) {
+	if scope == "" {
+		return false, nil
+	}
+
+	if !ValidateScope(scope) {
+		logger.Error(
+			"invalid scope",
+			"scope", scope,
+			"reason", "scopes should not contain meta-characters like * or ?, except in the last position",
```

Comment:
Is this visible at all to the user? Either way, "should" or "should not" is ambiguous. How about, "Unless in the last position, a scope cannot contain meta-characters such as `*` and `?`.

Reply:
nm, looks like you are following the SHOULD, MUST, etc. standard.